### ISSUE TITLE
Add eslint rule to avoid updating a ref in the render method

### ISSUE THREAD

**Issue by michenly** [Shopify/web-configs#229]
## Overview

The pattern come from these two issues:
https://github.com/Shopify/quilt/pull/1813
https://github.com/facebook/react/pull/18545

**TL;DR**
React core team is in discussion on adding warming for this and we can stop the bleeding earlier by adding this as a lint rule.

🚫 fail lint:
mutate refs in render method.

✅ pass lint:
[mutate ref in a useEffect ](https://github.com/Shopify/quilt/blob/main/packages/react-hooks/src/hooks/previous.ts#L3-L10)

## Type

- [x] New feature
- [ ] Changes to existing features

## Motivation

> What inspired this feature request? What problems were you facing,
> or what else makes you think this should be included?

...

## Labels

- [ ] Add the `Type: Feature Request` label to this issue.

### Area

- [ ] Add any relevant `Area: <area>` labels to this issue

### Scope

- Is this issue related to a specific package?

  - [ ] Tag it with the `Package: <package_name>` label.

- Is it related to a new package?
  - [ ] Tag it with the `New package` Label

---

## Checklist

- [ ] Please delete the labels section before submitting your issue
- [ ] I have described this issue in a way that is actionable (if possible)
